pkg/http/handler/timeout.go

```diff
@@ -85,6 +90,11 @@ func (h *timeToFirstByteTimeoutHandler) ServeHTTP(w http.ResponseWriter, r *http
 			}
 			return
 		case <-timeout.C:
+			timeoutExpired = true
+			// Return the timer early as it's practically useless from here on. This
+			// prevents very long requests (longer than the timeout) from hogging the
+			// timers unnecessarily.
+			returnTimer(timeout, timeoutExpired)
 			if tw.timeoutAndWriteError(h.body) {
```

Comment:
if this returns false then I think we wont return from the method, in which case we'll loop again, and - I think - block on the channel from the timer that we've returned to the pool, which seems probably unsafe?

Reply:
Good point, premature optimization YAY.

Reply:
Fixed.